### ISSUE TITLE
Treemacs hungs on large projects

### ISSUE THREAD

**Issue by iocanel** [Alexander-Miller/treemacs#175]
Each time I try to open a project that uses golang and vendoring for dependencies (which means really large projects) then treemacs hungs.

An example project is https://github.com/fabric8io/kubernetes-model.

I tried creating a predicate for ignoring the vendor folder (if there is a glide.yml file in the same dir), but it doesn't seem to work.
    
      (defun golang-vendor-p (name path)
        (let ((vendor (and (equal "vendor" name) (file-exists-p (format "%s/glide.yml" (file-name-directory path))))))
          (message (format "%s - %s" path vendor))
          vendor)
        ) 

      (setq treemacs-ignored-file-predicates '(golang-vendor-p))
 
I see in the Messages buffer that the predicate is called for for some of the files but then hugs (without displaying anything the vendor folder).

**Comment by Alexander-Miller** [comment]
> An example project is https://github.com/fabric8io/kubernetes-model.

I've tried this project and had no issues. Large projects in general should be no problem for treemacs. The files will all be stored in subdirectories, unless you expand them they might as well not exist. Even if you display many files at once, in my own tests rendering 3000 nodes takes only roughly twice as long as 30. 

What you can try is signal SIGUSR2 to emacs when it hangs. That will make emacs cancel whatever it is doing and throw up a stack trace. That should show which part is bogging you down. And as usual make sure that you're using the newest version.

A suspicion I have is that there might be something wrong with the getting/parsing of the git status. To verify that everything's right on this front try running the following code: 
`(treemacs--git-status-parse-function (treemacs--git-status-process-function "/home/a/Documents/git/kubernetes-model"))`

The output should be a hash table that contains output based on `git status --porcelain`.

> (setq treemacs-ignored-file-predicates '(golang-vendor-p))

Careful with that, you're ovewriting the original predicate. Best use `add-to-list`.

**Comment by iocanel** [comment]
@Alexander-Miller: Thanks for the quick response!

Since, you can't reproduce I played a little bit more with my settings and it seems that I can ONLY reproduce it in extended git-mode.

The part that seems to get stuck is the python code checks the git status. I can see it in my process list waiting forever. I tried to run it manually and it seems that its output was 32M. If I redirect it to a file it runs really really fast, so I assume its the elisp code that takes too long to process its output?

**Comment by Alexander-Miller** [comment]
The elisp does not do much in extended mode. The python outputs an alist that is read directly by the elisp interpreter. The elisp part just iterates over that alist and puts its content into a hash table. The output of the script should for example look like this - a bunch of cons cells mapping git status to a path.

```
❯❯❯ python -O treemacs-git-status.py /home/a/Documents/git/treemacs/src/scripts/
(("??" . "/home/a/Documents/git/treemacs/src/scripts/src/scripts/treemacs-dir-git-status.py")("M" . "/home/a/Documents/git/treemacs/src/scripts/src")("M" . "/home/a/Documents/git/treemacs/src/scripts/src/scripts"))⏎
```

You can see what the python outputs when run from elisp with this:
```
(-let*- [(repo "/home/a/Documents/git/kubernetes-model")
         (git-root (vc-call-backend 'Git 'root repo))
         (default-directory repo)
         (future (pfuture-new
                  treemacs-python-executable
                  "-O"
                  treemacs--git-status.py
                  (f-long git-root)))]
  (print (pfuture-await-to-finish future)))
```

If this gets stuck again we'll need to move on to print debugging the python.

**Comment by Alexander-Miller** [comment]
Closing due to inactivity.

**Comment by andrmuel** [comment]
Hello @Alexander-Miller 
It looks like I have bumped into the same issue. Executing your python command above produced **a lot** of output

    python -O path/to/treemacs/src/scripts/treemacs-git-status.py ~/path/to/repo/ | wc
          0   26325 1824238
This was caused by having a lot of files in the directory that were untracked and not in .gitignore (in my case a python virtualenv with a few modules). Adding the virtualenv directory to .gitignore helps, but I guess it would be nice if treemacs didn't get stuck even before adding it (in git status it only shows as one line, so its not super-obvious).

**Comment by Alexander-Miller** [comment]
> in git status it only shows as one line, so its not super-obvious

Try `git status --porcelain --ignored`. That'll show you what treemacs is working with. 

By its nature this issue cannot be solved smoothly, either way if you want the full info you *must* wait for emacs to read the output of the process, one way or another. All I can really do is decide which parts to drop. For now I've add `treemacs-max-git-entries` with a default value of 5000. That should take care of the hangups, but you might lose info you wouldn't want to lose.

Another thing I can think of is another option that lets you append something to the git command treemacs starts. You could set it to something like "| grep virtualenv -v". That could even work as a dir local variable so it won't affect all your projects.

**Comment by andrmuel** [comment]
Hmm, `git status --porcelain --ignored` wouldn't explain the behaviour; e.g. when I do `mkdir .foo && touch .foo/foo && touch .foo/bar && touch .foo/baz` in a git repo, `git status --porcelain --ignored` will still only show one entry for `?? .foo/`. 

For me personally, it doesn't matter much; after adding the virtualenv directory to .gitignore, everything works fine. I guess having a limit to 5000 entries makes sense. Even just a warning that things may be slow might be useful.

**Comment by Alexander-Miller** [comment]
You're right, if a directory is untracked it's content is not shown by git. In fact it looks like it's all my fault, I've explicitly made sure that the python script recursively include every single item inside an untracked directory.

And now that I think about it I can also remember why: Say you expand foo's parent node and then foo. Then you close the parent and open it again. foo will then be restored to its open state as well, so I need al that extra info so foo/foo & bar & baz are highlighted correctly. Same goes for recursively opening nodes. So yeah, that part needs to stay.

Can you revert your changes and check with the profiler exactly which part is the one that's blocking emacs? Based on my own simple test just letting the interpreter read the output into an alist (which can't be interrupted) is fast enough, reading some 50000 pairs takes 9ms. So the slow part must be something else, so I might be able to at least prevent things from hogging the UI. I played around with some macros today and you can basically get emacs to do work while there's no input with something like this: 

```
(progn 
  (setq x (number-sequence 0 10000000)) t)

(defun foo () 
  (while-no-input
    (while x
      (message "%s" (pop x))))
  (when x (run-with-idle-timer 1 nil #'foo)))
(foo)
;; to cancel:
;; (setq x nil)
```

**Comment by andrmuel** [comment]
I can't fully reproduce it (I'm on a different PC right now). Emacs still blocks, but only 2-3 seconds.

I was using the repository `https://github.com/zephyrproject-rtos/zephyr`. Steps to reproduce would be

    git clone https://github.com/zephyrproject-rtos/zephyr.git
    cd zephyr
    virtualenv --python=python2.7 .venv
    source .venv/bin/activate
    pip install -r scripts/requirements.txt

Then opening the project in treemacs.

The profiler just shows a lot of garbage collection going on.

    - ...                                                            3838  94%
       Automatic GC                                                  3838  94%
    + command-execute                                                 159   3%
    + timer-event-handler                                              80   1%
    + redisplay_internal (C function)                                   3   0%
      pfuture--append-output                                            1   0%

**Comment by Alexander-Miller** [comment]
I've now added `treemacs-git-command-pipe` so you can grep filter extreme cases like yours.

I'll leave the issue open as I experiment with my non-blocking task idea.
